lib/maths/CDataFrameUtils.cc

```diff
@@ -14,20 +14,52 @@
 #include <maths/CBasicStatistics.h>
 #include <maths/CMathsFuncs.h>
 #include <maths/CMic.h>
-#include <maths/CPRNG.h>
+#include <maths/COrderings.h>
 #include <maths/CQuantileSketch.h>
 #include <maths/CSampling.h>
 
+#include <boost/unordered_set.hpp>
+
 #include <vector>
 
 namespace ml {
 namespace maths {
-using TRowItr = core::CDataFrame::TRowItr;
 namespace {
+using TFloatVec = std::vector<CFloatStorage>;
+using TFloatVecVec = std::vector<TFloatVec>;
+using TFloatFloatPr = std::pair<CFloatStorage, CFloatStorage>;
+using TFloatFloatPrVec = std::vector<TFloatFloatPr>;
+using TFloatUSet = boost::unordered_set<CFloatStorage, std::hash<double>>;
+using TRowItr = core::CDataFrame::TRowItr;
+using TRowRef = core::CDataFrame::TRowRef;
+using TRowSampler = CSampling::CRandomStreamSampler<TRowRef>;
+
+//! Get a row column sampler.
+auto onSample(std::size_t i, std::size_t targetColumn, TFloatFloatPrVec& samples) {
+    return [i, targetColumn, &samples](std::size_t slot, const TRowRef& row) {
+        if (slot >= samples.size()) {
+            samples.resize(slot + 1, {0.0, 0.0});
+        }
+        samples[slot].first = row[i];
+        samples[slot].second = row[targetColumn];
+    };
+}
+
+//! Get a row sampler.
+auto onSample(TFloatVecVec& samples) {
+    return [&samples](std::size_t slot, const TRowRef& row) {
+        if (slot >= samples.size()) {
+            samples.resize(slot + 1, TFloatVec(row.numberColumns()));
+        }
+        row.copyTo(samples[slot].begin());
```

Comment:
This is sampling the whole row, to avoid having to repeatedly visit the data frame. Here we are completely overwriting the row which was previously in the slot. We use this method to get a very nearly random sample the rows in the data frame with a fixed size in a single pass over them, see `CRandomStreamSampler` for more detail.